VAGOV-TEAM-101466: Layout Page

ryguyk opened the pull request:
## Description
This work implements the bulk of the Layout page in Form Builder.
- The general structure is implemented, as shown in the screenshot below.
- Logic to calculate and display step statuses is added.
- A link to the Form Info page is added.
- Other links are left blank for now, as the pages do not exist.

To accomplish this work, some foundational pieces have also been added and/or refactored:
- A `DigitalForm` class is added that serves as a wrapper around a `digital_form` node. This allows business logic around step statuses and other related ideas to be called on the DigitalForm (node-wrapper) itself. Initially during this work, I was adding a lot of this code to the `DigitalFormsService` class, which effectively is a utility class that was requiring a `digital_form` node to be passed in to each call. 
 
This started to feel clunky:
```
'status' => $this->digitalFormsService->getDigitalFormStepStatus($this->digitalFormNode, 'contact_info'),
```
This is much better:
```
'status' => $this->digitalForm->getStepStatus('contact_info'),
```
- The service still exists, and its chief responsibility is fetching `digital_form` nodes and returning `DigitalForm` objects.

Relates to https://github.com/department-of-veterans-affairs/va.gov-team/issues/101466

## Testing done
- Unit/Functional tests added and updated.
- Manual testing done on all Form Builder pages, as this PR refactors some code that touches all pages.

## Screenshots
![image](https://github.com/user-attachments/assets/e9442884-300a-4380-98b5-90d829811ce9)


## QA steps
As an admin:
1. Navigate to `/form-builder`.
   - [x] Validate that you are redirected to `/form-builder/home`.
   - [x] On Home page, click on "Build a form".
   - [x] Validate that you are redirected to `/form-builder/form-info` (Form Info page in "create" mode).
1. On the Form Info page:
   - [x] Fill in the fields to create a new form.
   - [x] Click "Save and continue".
   - [x] Validate that you are redirected to the Layout page at `/form-builder/{nid}/layout`.
1. On the Layout page:
   - [x] Validate that the "Form Info" step is "COMPLETE" (green status label).
   - [x] Validate that the Introduction page and Form Steps are all "INCOMPLETE" (red status label). Note: Your personal information will eventually be complete by default, but that work still needs to be done.
   - [x] Validate that the "Review and sign" step is "COMPLETE" (green status label). Note: this page hasn't been completed yet in Form Builder, but this step is automatic. There is no database/content-model component to this.
   - [x] Click on the "View form info" button and validate that you are taken back to the Form Info page, this time in "edit" mode (i.e. fields are populated with the information you just entered to create this form).
   - [x] On the Form Info page, take note of the node id in the URL. Copy it, and navigate to the admin/default node-edit screen for the node in question `/node/{nid}/edit`.
1. On the default/admin node-edit screen:
   - [x] Add an "Address information" paragraph/chapter/step and save the node.
   - [x] Navigate back to `/form-builder/{nid}/layout` (using the same nid).
1. On the Layout page:
   - [x] Validate that "Step 2: Address information" is now "COMPLETE" (green status label).
1. Return to the default/admin node-edit screen (`node/{nid}/edit`), and on that page:
   - [x] Add a "List and loop" paragraph/chapter/step and save the node.
   - [x] Navigate back to `/form-builder/{nid}/layout`.
1. On the Layout page:
   - [x] Validate that there is now a "Step 4" showing the newly added List and loop with whatever title you entered.
   - [x] Validate that the link in this section reads "View {title-in-lowercase}"

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [x] Tests have been added if necessary.
- [x] Automated tests have passed.
- [x] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.
- [ ] If there are field changes, front end output has been thoroughly checked.

### Select Team for PR review

- [ ] `CMS Team`
- [ ] `Public websites`
- [ ] `Facilities`
- [ ] `User support`
- [ ] `Accelerated Publishing`
- [ ] `Form Engine`

### Is this PR blocked by another PR?

- [ ] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing announcement

Is an announcement needed to let editors know of this change?
- [ ] Yes, and it's written in issue ____ and queued for publication.
  - [ ] Merge and ping the UX writer so they are ready to publish after deployment
- [ ] Yes, but it hasn't yet been written
  - [ ] Don't merge yet -- ping the UX writer to write and queue content
- [x] No announcement is needed for this code change.
  - [ ] Merge & carry on unburdened by announcements

ryguyk commented on docroot/modules/custom/va_gov_form_builder/src/EntityWrapper/DigitalForm.php:
```diff
@@ -0,0 +1,206 @@
+/**
+ * A wrapper class around Digital Form nodes.
+ *
+ * @method int id()
+ * @method string getTitle()
+ * @method \Drupal\Core\Field\FieldItemListInterface get(string $field_name)
+ * @method \Symfony\Component\Validator\ConstraintViolationListInterface validate()
+ * @method int save() Saves the entity and returns the save status
```
See the code comment on the `__call` magic method. 

The purpose of this class is to wrap a `digital_form` node so that we can add business logic to it (e.g. status calculations) and encapsulate that logic with the node itself. In doing so, we still want to be able to make all the calls we were previously making on the nodes, so we implement the `__call` magic method to forward these calls to the underlying node without having to implement each of them explicitly. This works well, except that PHPStan doesn't like it (calls are dynamic, and the static analyzer doesn't think they exist). This annotation here simply lists those without needing to re-implement, which satisfies PHPStan. I'd rather not have to do this, because we'll need to add to this with each new method we call on the underlying node, but it's a smaller price than having to implement it directly (though that wouldn't be difficult either, as each of those methods would effectively just be a single line that amounts to calling the method on the `$node` private variable). This approach feels a little lighter weight, though.

ryguyk commented on docroot/modules/custom/va_gov_form_builder/src/Form/Base/FormBuilderBase.php:
```diff
@@ -107,58 +117,27 @@ protected function getDigitalFormNodeFieldValue($fieldName) {
     return $form;
   }
 
-  /**
-   * Determines if `digitalFormNode` has a chapter (paragraph) of a given type.
-   *
-   * @param string $type
-   *   The chapter (paragraph) type.
-   *
-   * @return bool
-   *   TRUE if the chapter exists; FALSE if the chapter
-   *   does not exist or the node does not exist.
-   */
-  protected function digitalFormNodeHasChapterOfType($type) {
-    if (empty($this->digitalFormNode)) {
-      return FALSE;
-    }
-
-    $chapters = $this->digitalFormNode->get('field_chapters')->getValue();
-
-    foreach ($chapters as $chapter) {
-      if (isset($chapter['target_id'])) {
-        $paragraph = $this->entityTypeManager->getStorage('paragraph')->load($chapter['target_id']);
-        if ($paragraph) {
-          if ($paragraph->bundle() === $type) {
-            return TRUE;
-          }
-        }
-      }
-    }
-
-    return FALSE;
-  }
-
```
This got moved to the `DigitalForm` class.

derekhouck commented:
> This started to feel clunky:
> ```'status' => $this->digitalFormsService->getDigitalFormStepStatus($this->digitalFormNode, 'contact_info'),```
> This is much better:
> ```'status' => $this->digitalForm->getStepStatus('contact_info'),```

I agree. Good change.

derekhouck reviewed (changes requested):
The layout page is currently cached and will not update to show changes to the node until the cache is flushed. This could create a confusing user experience.

ryguyk commented:
> The layout page is currently cached and will not update to show changes to the node until the cache is flushed. This could create a confusing user experience.

I removed caching on Form Builder pages. For now, this is fine. I added a `@todo` to possibly address more granular or selective caching in the future, but for now this is a reasonable solution.

derekhouck commented on docroot/modules/custom/va_gov_form_builder/src/EntityWrapper/DigitalForm.php:
This feels like a very important file for maintainability and ease of implementation moving forward.

derekhouck reviewed (approved):
LGTM

ryguyk commented on docroot/modules/custom/va_gov_form_builder/src/EntityWrapper/DigitalForm.php:
```diff
@@ -0,0 +1,206 @@
+   *   or if the step name does not exist.
+   */
+  public function getStepStatus($stepName) {
+    if ($stepName === 'form_info') {
+      // If the node exists, this will necessarily be complete.
+      return 'complete';
+    }
+
+    if ($stepName === 'review_and_sign') {
+      // This is added automatically by the Forms Library.
+      return 'complete';
+    }
+
+    if (in_array($stepName, [
+      'intro',
+      'confirmation',
+    ])) {
+      // These haven't been handled yet.
+      // Return 'incomplete' for the time being.
+      return 'incomplete';
+    }
+
+    // Standard steps are complete if a corresponding chapter exists.
+    if (array_key_exists($stepName, self::STANDARD_STEPS)) {
+      $paragraphName = self::STANDARD_STEPS[$stepName];
+      return $this->hasChapterOfType($paragraphName)
+        ? 'complete'
+        : 'incomplete';
+    }
+
+    return 'incomplete';
```
I like this thought. I'm not ready to implement it yet, as I'm not sure this is complete enough for optimization, but I'll keep this in mind as I continue to work with this function.
